pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java

```diff
@@ -153,13 +155,18 @@ public void resetCloseFuture() {
         PulsarApi.MessageMetadata metadata = Commands.parseMessageMetadata(metadataAndPayload);
 
         try {
-            if (metadata.hasNumMessagesInBatch()) {
+            if (!metadata.hasOrderingKey() && !metadata.hasPartitionKey() && metadata.hasNumMessagesInBatch()) {
                 // If the message was part of a batch (eg: a batch of 1 message), we need
                 // to read the key from the first single-message-metadata entry
+                PulsarApi.CompressionType compressionType = metadata.getCompression();
+                CompressionCodec codec = CompressionCodecProvider.getCompressionCodec(compressionType);
```

Comment:
We should avoid decompressing in the broker in any case. This is very dangerous from a CPU usage point of view.

Reply:
I'm not sure how to handle #7107 that fixed. Maybe we can use the first single message key as the batch message key, but this requires all clients catchup.

So, I think we can onboard this PR first ensure correctness, and I will create an issue for the clients catchup work.

And I have add check `if (!metadata.hasOrderingKey() && !metadata.hasPartitionKey() && metadata.hasNumMessagesInBatch())`, if the clients set the key of the batch message to the key of the first message, it can reduce overhead.

I have test with the KeyBasedBatcher. It already set the batch message key.

Reply:
A better way would be to ensure that producers (in every lang) are attaching the routing key on the message metadata protobuf (at the batch level). That is not compressed so we should always be safe.

Reply:
Make sense, so it's better to revert #7107 and other lang clients should attach the routing key on the message metadata.

Reply:
@merlimat I have addressed your comment, please take a look, thanks.